File: Python3/Sample.py
"""sample management"""
import numpy as np

class Sample:
    def __init__(self, x, options):
        # input dimension
        self.n = len(x)

        # initial sample
        Y = x + options['tr_delta'] * np.eye(self.n)
        self.Y = np.concatenate((Y, x.reshape(1,self.n)))

        # function value of initial sample
        self.fY = np.empty(len(self.Y))
        self.fY[:] = np.nan

    @property
    def m(self):
        m = len(self.Y)
        assert m == len(self.fY), \
            'Number of sample points and number of function values do not match!'
        return m

    def Ycentered(self, center):
        return self.Y - center

    def distance(self, center):
        return np.linalg.norm(self.Ycentered(center), axis=1)

    def addpoint(self, point):
        # assert self.m < options['sample_max'],
        #     'Attemp to add more points when the sample set is already at its capacity. '
        self.Y = np.vstack((self.Y, point))
        self.fY = np.append(self.fY, [np.nan])
    
    def auto_delete(self, model, options):
        """ Delete points that are too far away from the trust region."""
        # After the deletion, the sample size should be no less than
        #  sample_min, and no more than sample_max.

        distance = self.distance(model.center)

        # indeces of the farthest ones; 
        # We should not delete too many points so that sample size goes under sample_min. 
        farthest = np.argsort(distance)[-(self.m - options['sample_min']):]
        farthest = np.array(farthest)
        # indeces of the ones that are too far
        toofar = np.where(distance > options['sample_toremove'] * model.delta)[0]
        # indeces of the points to be removed
        if toofar.shape[0] == 0 and self.m > options['sample_max']:
            # We need to keep the sample size under its maximum. 
            toremove = farthest[-(self.m - options['sample_max']):]
        else:
            toremove = np.intersect1d(farthest, toofar)

        # remove points
        self.Y = np.delete(self.Y, toremove, axis=0)
        self.fY = np.delete(self.fY, toremove)


def quadratic_terms(x):
    n = len(x)
    x = x[np.newaxis] # turn x from 1D to a 2D array
    temp = x * x.T - (1 - 1/np.sqrt(2)) * np.diag(pow(x, 2)[0]) 
    
    return temp[np.tril_indices(n)]

File: MATLAB/Sample.m
classdef Sample < handle
% sample set management
% 
% Written by Liyuan Cao @Lehigh University in 2021.

properties
    n % input dimension
    Y % sample set in the input space (Each row is a point in R^n.)
    fY % sample set function values
end

properties (Dependent)
    m % sample size 
end

methods
    
    function self = Sample(x, options)
        % input dimension
        self.n = length(x);

        % initial sample
        Y = x + options.tr_delta * eye(self.n);
        self.Y = [Y; x'];

        % function value of initial sample
        self.fY = nan(size(self.Y, 1), 1);
    end
    
    function o = get.m(self)
        o = size(self.Y, 1);
        if o ~= length(self.fY)
            error('Number of sample points and number of function values do not match!')
        end
    end

    function value = Ycentered(self, center)
        % Y after the origin is shifted to a specified center 
        value = self.Y - center';
    end

    function value = distance(self, center)
        % Euclidean distance from each sample point to a specified center 
        value = vecnorm(self.Ycentered(center), 2, 2);
    end

    function addpoint(self, point)
%         if self.m >= options.sample_max
%             error('Attemp to add more points when the sample set is already at its capacity. ')
%         end
        self.Y = [self.Y; point'];
        self.fY = [self.fY; nan];
    end
    
    function auto_delete(self, model, options)
        % Delete points that are too far away from the trust region.
        % After the deletion, the sample size should be no less than
        %  sample_min, and no more than sample_max.

        distance = self.distance(model.center);

        % indeces of the ones that are too far
        toofar = find(distance > options.sample_toremove * model.delta);
        % indeces of the farthest ones
        % We should not delete too many points so that sample size goes under sample_min. 
        [~,farthest] = maxk(distance, max(self.m - options.sample_min,0));
        % indeces of the points to be removed
        if isempty(toofar) && self.m > options.sample_max
            % We need to keep the sample size under its maximum. 
            toremove = farthest(1:self.m-options.sample_max);
        else
            toremove = intersect(farthest, toofar);
        end

        % remove points
        self.Y(toremove,:) = [];
        self.fY(toremove) = [];
    end
end
end


function output = quadratic_terms(x)
    n = length(x);
    temp = x * x' - (1 - 1/sqrt(2)) * diag(x.^2);
    output = temp(tril(ones(n)));
end